// aa-conla-utils/dist/src/types/factories/@openzeppelin/contracts/utils/Create2__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Create2__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "Create2EmptyBytecode",
        type: "error",
    },
    {
        inputs: [],
        name: "Create2FailedDeployment",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "balance",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "needed",
                type: "uint256",
            },
        ],
        name: "Create2InsufficientBalance",
        type: "error",
    },
];
const _bytecode = "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212207ee53a45e64b2ab5a3569d84fe3057c9111b0c47e224fcc04098bc2550fda95364736f6c63430008170033";
const isSuperArgs = (xs) => xs.length > 1;
class Create2__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.Create2__factory = Create2__factory;
Create2__factory.bytecode = _bytecode;
Create2__factory.abi = _abi;
